src/redprl/new_machine.sml

```diff
@@ -0,0 +1,306 @@
+functor NewMachine (Sig : MINI_SIGNATURE) :
+sig
+  type sign = Sig.sign
+  type abt = RedPrlAbt.abt
+  type 'a machine
+
+  datatype stability = 
+     CUBICAL
+   | NOMINAL
+
+  datatype blocker =
+     VAR of RedPrlAbt.variable
+   | METAVAR of RedPrlAbt.metavariable
+
+  exception Neutral of blocker
+  exception Unstable
+
+  val step : sign -> stability -> abt machine -> abt machine
+end = 
+struct
+  structure Tm = RedPrlAbt
+  structure Syn = Syntax
+  
+
+  type sign = Sig.sign
+
+  datatype 'a closure = <: of 'a * environment
+  withtype environment = Tm.abt closure Tm.bview Metavar.Ctx.dict * Tm.abt closure Var.Ctx.dict * Tm.param Sym.Ctx.dict
+
+  fun @@ (f, x) = f x
+  infixr @@
+
+
+  infix 6 <:
+  infix 3 ||
+
+  open Tm infix 7 $ $$ $# infix 6 \
+  structure O = RedPrlOpData
+  structure P = struct open RedPrlParameterTerm RedPrlSortData end
+
+  datatype hole = HOLE
+  datatype continuation =
+     APP of hole * abt
+   | HCOM of symbol O.dir * hole * abt * (symbol O.equation * (symbol * abt)) list
+   | COE of symbol O.dir * (symbol * hole) * abt
+   | FST of hole
+   | SND of hole
+
+  type frame = continuation closure
+  type stack = frame list
+
+  datatype 'a machine = || of 'a closure * stack
+
+
+  datatype stability = 
+     CUBICAL
+   | NOMINAL
+
+  datatype blocker =
+     VAR of variable
+   | METAVAR of metavariable
+
+  exception Neutral of blocker
+  exception Unstable
+
+  val todo = Fail "TODO"
+  fun ?e = raise e
+
+  fun lookupSym psi x = 
+    Sym.Ctx.lookup psi x
+    handle Sym.Ctx.Absent => P.ret x
+
+  fun readParam psi : param -> param = 
+    P.bind (lookupSym psi)
+
+  fun insertVar x cl (mrho, rho, psi) = 
+    (mrho, Var.Ctx.insert rho x cl, psi)
+
+  fun insertMeta meta bcl (mrho, rho, psi) = 
+    (Metavar.Ctx.insert mrho meta bcl, rho, psi)
+
+  fun insertSym u r (mrho, rho, psi) = 
+    (mrho, rho, Sym.Ctx.insert psi u r)
+
+  fun dimensionsEqual (_, _, psi) (r1, r2) = 
+    P.eq Sym.eq (readParam psi r1, readParam psi r2)
+
+  fun findTrueEquationIndex env = 
+    let
+      fun aux i [] = NONE
+        | aux i ((r,r') :: eqs) =
+          if dimensionsEqual env (r, r') then 
+            SOME i
+          else 
+            aux (i + 1) eqs
+    in
+      aux 0
+    end
+
+  fun stepView sign stability = 
+    fn `x <: (mrho, rho, psi) || stk =>
+       (Var.Ctx.lookup rho x || stk
+        handle Var.Ctx.Absent => raise Neutral (VAR x))
+     | meta $# (ps, ms) <: env || stk =>
+       let
+         val (mrho, rho, psi) = env
+         val (us, xs) \ (m <: (mrho', rho', psi')) = Metavar.Ctx.lookup mrho meta handle Metavar.Ctx.Absent => raise Neutral (METAVAR meta)
+         val rho'' = ListPair.foldl (fn (x, n, rho) => Var.Ctx.insert rho x (n <: env)) rho' (xs, ms)
+         val psi'' = ListPair.foldl (fn (u, (r, _), psi) => Sym.Ctx.insert psi u (readParam psi r)) psi' (us, ps)
```

Comment:
PS: My other ideas, if working, will only need to force `ms`, which might be better because it seems all the `xs` or `ms` generated by the machine would be empty. That is, it seems the only non-empty `xs` (or `ms`) are given by the users.